### ISSUE TITLE
Various `tasks.json` issues caused by v1.23.0

### ISSUE THREAD

**Issue by eladavron** [microsoft/vscode-docker#3722]
As soon as I updated to v1.23.0 several things broke:

* Environment variables in `tasks.json` are no longer being expanded
* `dockerRun -> customOptions` now treats `--user foo` as one string instead of two, adding an unnecessary `\` after `--user` causing Docker to not understand the query 
(Workaround for that was easy, just convert it to `--user=foo` but still - a warning would have been nice)

**Comment by bwateratmsft** [comment]
For the environment variables, do you mean the [`${env:FOO}` syntax](https://code.visualstudio.com/docs/editor/variables-reference#_environment-variables) inside a task definition, or something else?

**Comment by narnold113** [comment]
This is also happening in dockerRun -> command. For example,
`"command": "node --inspect=0.0.0.0 lib/docker-local/entry_point.js",`
yields:
`node\ --inspect=0.0.0.0\ lib/docker-local/entry_point.js`

It escapes the whitespace causing an error

**Comment by eladavron** [comment]
> For the environment variables, do you mean the [`${env:FOO}` syntax](https://code.visualstudio.com/docs/editor/variables-reference#_environment-variables) inside a task definition, or something else? I tested the following (with `DJANGO_IMAGE_TAG=1`) and it worked as expected:
> 
> ```json
> {
> 	"type": "docker-build",
> 	"label": "docker-build",
> 	"platform": "python",
> 	"dockerBuild": {
> 		"tag": "djangotest:${env:DJANGO_IMAGE_TAG}",
> 		"dockerfile": "${workspaceFolder}/Dockerfile",
> 		"context": "${workspaceFolder}",
> 		"pull": true
> 	},
> },
> ```

Using `${env:env_name}` instead of `${env_name}` causes it to use the Windows Host's environment variable, not WSL's, which is somehow even weirder.

**Comment by bwateratmsft** [comment]
Ah, I see. Can you give an example where it worked in the past? I wonder if we're escaping something that we weren't in the past...

**Comment by bwateratmsft** [comment]
(Just for clarity on the label/milestone changes, the fix for the `customOptions` / `command` being over-escaped is going to go out in 1.23.1 today--I reopened this issue to cover the environment variable issue)

**Comment by bwateratmsft** [comment]
A fix for the `customOptions` / `command` overescaping issue has now been released in Docker extension version 1.23.1.

**Comment by eladavron** [comment]
> Ah, I see. Can you give an example where it worked in the past? I wonder if we're escaping something that we weren't in the past...

Yes.
Let's say in WSL I have an environment variable `FOO_PATH=/path/in/wsl`
Then in my `tasks.json` I'd use:
```json
{
    "tasks": [
        {
            "type": "docker-run",
            "label": "docker-run: debug",
            "dockerRun": {
                "volumes": [
                    {
                        "containerPath": "/path/in/container",
                        "localPath": "${FOO_PATH}"
                    }
                ]
            }
        }
    ]
}
```

And then when the container ran, it would mount whatever path was defined in `FOO_PATH` (in this case `/path/in/wsl`) to the container's `/path/in/container`

Since 1.23.0, it's no longer resolved, and instead tries to map the path called `$FOO_PATH` (here's the command it uses in the terminal - edited for brevity):
```
docker container run --mount 'type=bind,source=${FOO_PATH},destination=/path/in/container' 
```
Whereas before it would look like:
```
docker container run --mount 'type=bind,source=/path/in/wsl,destination=/path/in/container' 
```

**Comment by bwateratmsft** [comment]
Are you using "Remote - WSL" to connect to the WSL instance, or just using `wsl.exe` as the shell?

**Comment by eladavron** [comment]
> Are you using "Remote - WSL" to connect to the WSL instance, or just using `wsl.exe` as the shell for VSCode?

I am connected to the WSL instance using the "Remote - WSL" extension

**Comment by bwateratmsft** [comment]
Strange, I would definitely not have expected the `${env:FOO}` variables to be picked up from the Windows host in that case. We're looking at no longer using a shell to execute these commands, which would definitely break the inline environment variables, leaving only `${env:FOO}` as an option, but if it's choosing the value from the Windows host that seems like wrong behavior to me.

**Comment by bwateratmsft** [comment]
@meganrogge we use `CustomExecution` tasks for `docker-build` / `docker-run` / etc., using the `resolveDefinition` that is passed into the callback. Is it expected that it would use the Windows host's environment variable values instead of the WSL "machine"'s variables?

**Comment by meganrogge** [comment]
Yes, that looks like a bug here https://github.com/microsoft/vscode/blob/7df4fd594f047d2a6c8b884d6bfb985d6e2a468c/src/vs/workbench/contrib/tasks/browser/terminalTaskSystem.ts#L1075-L1086

**Comment by eladavron** [comment]
Right thing or not - functionality is broken or at the very least regressed here.
Is there not going to be a way to use WSL environment variables? It's useful when several users are using the same workspace cinfig across different machines.

**Comment by bwateratmsft** [comment]
@eladavron depending on where your `/path/in/wsl` is, there might be [other variables](https://code.visualstudio.com/docs/editor/variables-reference) that can help. For example, I tried using `${userHome}`, and it substituted the correct path in (the one in WSL). Another thing I tried that worked successfully was using the [`config` substitution](https://code.visualstudio.com/docs/editor/variables-reference#_configuration-variables)--you can set a variable in the settings specifically in WSL (`Preferences: Open Remote Settings (JSON) (WSL: <distroname>)` in the command palette), or in the workspace (`.vscode/settings.json`). The setting does not need to be a "real" setting, i.e. you can name it anything you want. The setting will show as faded because VSCode doesn't associate it with VSCode itself or any extension, but it works just fine.

Example with `${userHome}`:
```json
{
    "type": "docker-run",
    "label": "docker-run: debug",
    "dependsOn": [
        "docker-build: debug"
    ],
    "dockerRun": {
        "volumes": [
            {
                "localPath": "${userHome}",
                "containerPath": "/home/from/wsl",
                "permissions": "ro"
            }
        ]
    },
    "netCore": {
        "appProject": "${workspaceFolder}/dotnetproj.csproj",
        "enableDebugging": true
    }
},
```
Resulting command line (truncated for brevity): `docker container run --detach --tty --name 'dotnetproj-dev' --mount 'type=bind,source=/home/bwater,destination=/home/from/wsl,readonly' dotnetproj:dev`

Example with `${config:foo.bar}`:
![image](https://user-images.githubusercontent.com/36966225/205677959-a5810ab0-97b7-47a9-bbe5-5fac049386e7.png)
```json
{
    "type": "docker-build",
    "label": "docker-build: debug",
    "dependsOn": [
        "build"
    ],
    "dockerBuild": {
        "tag": "dotnetproj:${config:foo.bar}",
        "target": "base",
        "dockerfile": "${workspaceFolder}/Dockerfile",
        "context": "${workspaceFolder}",
        "pull": true
    },
    "netCore": {
        "appProject": "${workspaceFolder}/dotnetproj.csproj"
    }
},
```
Resulting command line: `docker image build --pull --file '/home/bwater/src/dotnetproj/Dockerfile' --target 'base' --tag 'dotnetproj:baz' --label 'com.microsoft.created-by=visual-studio-code' '/home/bwater/src/dotnetproj'`

**Comment by bwateratmsft** [comment]
I'm going to close this issue as fixed in 1.23.1--referring to the overescaping of the `customOptions` and `command` properties on tasks.

The other issue, about environment variables, looks to be [external](https://github.com/microsoft/vscode/issues/167969). As for why the older `${FOO}` syntax that used to be working but no longer is--I'm pretty sure it's because of one or both of two reasons:
1. The old code used the `-v` flag, whereas the new code uses the `--mount` flag.
2. The old code only did simple quoting of arguments, whereas the new code does a little bit more sophisticated quoting and escaping.

I know it's not a very satisfying answer, but this way of using environment variables was never tested nor intended to work--rather, the `${env:FOO}` syntax that VSCode provides should be used. But, as noted, it's [broken](https://github.com/microsoft/vscode/issues/167969) in WSL. Hopefully it will be fixed soon, and in the meantime, the workarounds mentioned [here](https://github.com/microsoft/vscode-docker/issues/3722#issuecomment-1337593979) may be useful.

**Comment by eladavron** [comment]
This workaround isn't perfect - the whole idea of using an environment variable is that all my devs can have the same config and don't need to manually adjust their own, but I guess I'm out of luck and will simply have to tell them to all make sure their target folders are in the same path relative to their home dir.
I'll keep my eyes on that other thread though, thanks for linking it.
